Source/WebKit/WebProcess/Storage/WebSWClientConnection.cpp

```diff
@@ -410,25 +410,35 @@ void WebSWClientConnection::notifyRecordResponseBodyEnd(RetrieveRecordResponseBo
         callback(makeUnexpected(WTFMove(error)));
 }
 
+static RefPtr<Page> pageFromScriptExecutionContextIdentifier(ScriptExecutionContextIdentifier clientIdentifier)
+{
+    RefPtr document = Document::allDocumentsMap().get(clientIdentifier);
+    if (!document) {
+        RefPtr loader = DocumentLoader::fromScriptExecutionContextIdentifier(clientIdentifier);
+        RefPtr frame = loader ? loader->frame() : nullptr;
+        return frame ? frame->page() : nullptr;
+    }
+    return document->page();
+}
+
 void WebSWClientConnection::focusServiceWorkerClient(ScriptExecutionContextIdentifier clientIdentifier, CompletionHandler<void(std::optional<ServiceWorkerClientData>&&)>&& callback)
 {
-    RefPtr client = Document::allDocumentsMap().get(clientIdentifier);
-    auto* page = client ? client->page() : nullptr;
+    RefPtr page = pageFromScriptExecutionContextIdentifier(clientIdentifier);
     if (!page) {
         callback({ });
         return;
     }
 
     WebPage::fromCorePage(*page)->sendWithAsyncReply(Messages::WebPageProxy::FocusFromServiceWorker { }, [clientIdentifier, callback = WTFMove(callback)]() mutable {
-        RefPtr client = Document::allDocumentsMap().get(clientIdentifier);
-        RefPtr frame = client ? client->frame() : nullptr;
-        auto* page = frame ? frame->page() : nullptr;
+        RefPtr document = Document::allDocumentsMap().get(clientIdentifier);
```

Comment:
I suppose there could still be an issue here if the DocumentLoader hasn't created the Document by the time this callback runs? Not sure what the consequences of that are.

Reply:
There is a compromise here, either we focus right away the tab even if we have not yet received the response.
Or we wait for the Document to proceed with the focus steps.
The spec is more inline with the latter.

Reply:
Maybe we should go to the UIProcess but wait for the Document creation to trigger these steps and resolve the promise.
I'll update the patch accordingly.

Reply:
Actually, I'll land it as is for now as it is an improvement.
I will work on waiting for the Document creation to resolve the focus promise in a follow-up PR.

Reply:
https://github.com/WebKit/WebKit/pull/33240